### ISSUE TITLE
About object loss function

### ISSUE THREAD

**Issue by SISTMrL** [ultralytics/yolov5#808]
## ❔Question

There three lines code relative to the caculation of object loss function in general.py:

- balance = [4.0, 1.0, 0.4] if np == 3 else [4.0, 1.0, 0.4, 0.1]  # P3-5 or P3-6
- tobj[b, a, gj, gi] = (1.0 - model.gr) + model.gr * giou.detach().clamp(0).type(tobj.dtype)  # giou ratio
- lobj += BCEobj(pi[..., 4], tobj) * balance[i]
- lobj *= h['obj'] * s * (1.4 if np == 4 else 1.)

I have three questions:
The first question is what's the meaning of balance and why in the third line code, lobj should be multiplied to balance[i]

The second question is the second line code, why considering the giou and the meaning of model.gr

The third question is about the hyparameters: I can't understand h['obj'], what's its effects and other two parameters: h['giou'] and h['cls']. In your code, you said is loss gain but i can't understand

The questions seem a bit too many, please kindly answer them patiently, thanks very much!!!!!!!!

**Comment by ZhengfengLee** [comment]
also want to figure out those questions and the followings:
`s = 3 / np  # output count scaling`
`lbox *= h['giou']` 
`lobj *= h['obj'] * s * (1.4 if np == 4 else 1)`
`lcls *= h['cls'] * s`
whats the meanings of the 's' (output count scaling) and scaling-factor '1.4' in second row?

**Comment by Edwardmark** [comment]
@glenn-jocher Could you explain about the balance?
@SISTMrL Did you find out what does balance mean? Please help me out, thanks.

**Comment by zfli-xdu** [comment]
just some weights of different detection layers for balance.

**Comment by Edwardmark** [comment]
> just some weights of different detection layers for balance.

Thanks.
